include/mbedtls/ssl.h

```diff
@@ -1099,6 +1101,14 @@ mbedtls_dtls_srtp_info;
 
 #endif /* MBEDTLS_SSL_DTLS_SRTP */
 
+/** Human-friendly representation of the (D)TLS protocol version. */
+typedef enum
+{
+    MBEDTLS_SSL_VERSION_UNKNOWN, /*!< Context not in use or version not yet negotiated. */
+    MBEDTLS_SSL_VERSION_TLS1_2 = 0x0303, /*!< (D)TLS 1.2 */
+    MBEDTLS_SSL_VERSION_TLS1_3 = 0x0304, /*!< (D)TLS 1.3 */
```

Comment:
> cons:
>    need for a translation in `mbedtls_ssl_write/read_version` for TLS

Encapsulated in `mbedtls_ssl_write/read_version`; choice can be deferred to some later time.

> pros:
>    less ambiguity for DTLS as we do not use TLS related values.

Encapsulated in `mbedtls_ssl_write/read_version`, except for those two existing pathological cases for which you requested special-casing; choice can be deferred to some later time.

>    when we save/restore session one byte is and will be enough to store the protocol version.

Encapsulated in `mbedtls_ssl_write/read_version`; choice can be deferred to some later time.

>    in the list of ciphersuites we can gain even more space. Only one byte per protocol version.

Structure alignment of `char *name` hides this change if the other changes I suggested to `mbedtls_ssl_ciphersuite_t` are made.

> I suggest 2 and 3 rather than 1 and 2 as it seems better aligned with TLS 1.2/3 but not completely sure if this is a good idea.

You **must not** change these from their existing values for the lower byte (3 and 4), or you will break existing serialized sessions that have been saved on disk.  (Since currently supported only for TLS 1.2 in mbedtls 3.0.0, the lower byte of MBEDTLS_SSL_VERSION_TLS1_2 must remain 3.)

I respectfully request that we move forward with this PR with the current values of `mbedtls_ssl_protocol_version`.  Once this large PR is merged, if there is a need to change the enum values, that can be done later, expanding the internal functions `mbedtls_ssl_write/read_version` at that time.

This PR remains useful, doing the heavy lifting of using a single enum for TLS protocol version, rather than major and minor, and replaces less clean code which looks only at the minor version (except for session serialization save/load).

Aside: as this is a large PR, I would like to go back and squash some commits after the review is complete and before merging.  One can `git fetch` and `git diff enum-tls-vers..gstrauss/enum-tls-vers` (assuming the remote is 'gstrauss') to confirm that the code changes are equivalent after the squash.

Reply:
>    in ssl_tls12_server.c, the check if( tls_version < 0x0300 )
> For those I would suggest some ad-hoc solutions. In the server case, just read the major version (one byte) and check it is equal to 3.

Won't that also need to special-case the raw DTLS version?

Reply:
> in ssl_tls12_client.c, the check if( tls_version < 0x0302 || /* TLSv1.1 */
> In the DTLS client case, read the "raw" on-wire protocol version with a MBEDTLS_GET_UINT16_BE, if it is equal to 0xfeff then accept it, if it is not equal to 0xfeff read the version with mbedtls_ssl_read_version, check this is a valid version and then compare it to the configured maximum version.

I have replaced the version check with something more forward-looking to DTLS 1.3.  mbedtls is guilty of "version intolerance" as described in the draft for DTLS 1.3: https://tlswg.org/dtls13-spec/draft-ietf-tls-dtls13.html#name-clienthello-message  For DTLS 1.3, the version will need to be parsed from the "supported_version" ClientHello extension.

Reply:
:rocket: proposed way forwards is to remove the two version negotiation checks (the "pathological cases"), and leave the enum as-is, and leave `mbedtls_ssl_write/read_version` as-is for this PR.

Reply:
> You must not change these from their existing values for the lower byte (3 and 4), or you will break existing serialized sessions that have been saved on disk. (Since currently supported only for TLS 1.2 in mbedtls 3.0.0, the lower byte of MBEDTLS_SSL_VERSION_TLS1_2 must remain 3.)

I do not think we support this. The version of the library is saved in serialized session headers and check against the version of the library that try to load them. Thus session (and context) saved with a previous version of the library cannot be use with a new version of the library.

Reply:
> rocket proposed way forwards is to remove the two version negotiation checks (the "pathological cases"), and leave the enum as-is, and leave `mbedtls_ssl_write/read_version` as-is for this PR.

Yes let's proceed that way.

Reply:
> > You must not change these from their existing values for the lower byte (3 and 4), or you will break existing serialized sessions that have been saved on disk. (Since currently supported only for TLS 1.2 in mbedtls 3.0.0, the lower byte of MBEDTLS_SSL_VERSION_TLS1_2 must remain 3.)
> 
> I do not think we support this. The version of the library is saved in serialized session headers and check against the version of the library that try to load them. Thus session (and context) saved with a previous version of the library cannot be use with a new version of the library.

When `ssl_session_save()` is called from `mbedtls_ssl_session_save()`, the `omit_header` flag is passed as false.
When `ssl_session_save()` is called from `mbedtls_ssl_context_save()`, the `omit_header` flag is passed as true.
I figured that it was safest to preserve the exact current behavior, which uses the value: 0x03.

In any case, this is not an issue for this PR.